File: Task/Additive-primes/C-sharp/additive-primes.cs
internal class Program
{
    private static void Main(string[] args)
    {
        long primeCandidate = 1;
        long additivePrimeCount = 0;
        Console.WriteLine("Additive Primes");

        while (primeCandidate < 500)
        {
            if (IsAdditivePrime(primeCandidate))
            {
                additivePrimeCount++;

                Console.Write($"{primeCandidate,-3} ");

                if (additivePrimeCount % 10 == 0)
                {
                    Console.WriteLine();
                }
            }

            primeCandidate++;
        }

        Console.WriteLine();
        Console.WriteLine($"Found {additivePrimeCount} additive primes less than 500");
    }

    private static bool IsAdditivePrime(long number)
    {
        if (IsPrime(number) && IsPrime(DigitSum(number)))
        {
            return true;
        }

        return false;
    }

    private static bool IsPrime(long number)
    {
        if (number < 2)
        {
            return false;
        }

        if (number % 2 == 0)
        {
            return number == 2;
        }

        if (number % 3 == 0)
        {
            return number == 3;
        }

        int delta = 2;
        long k = 5;

        while (k * k <= number)
        {
            if (number % k == 0)
            {
                return false;
            }

            k += delta;
            delta = 6 - delta;
        }

        return true;
    }

    private static long DigitSum(long n)
    {
        long sum = 0;

        while (n > 0)
        {
            sum += n % 10;
            n /= 10;
        }

        return sum;
    }
}

File: Task/Additive-primes/Zen-C/additive-primes.c
fn digit_sum(n: int) -> int {
    let sum = 0;
    while n > 0 {
        sum += n % 10;
        n /= 10;
    }
    return sum;
}

fn is_prime(n: int) -> bool {
    if n < 2      { return false; }
    if n % 2 == 0 { return n == 2; }
    if n % 3 == 0 { return n == 3; }
    let d = 5;
    while d * d <= n {
        if n % d == 0 { return false; }
        d += 2;
        if n % d == 0 { return false; }
        d += 4;
    }
    return true;
}

fn main() {
    println "Additive primes less than 500:";
    let i = 2;
    let count = 0;
    while i < 500 {
        if is_prime(i) && is_prime(digit_sum(i)) {
            print "{i:3d}  ";
            if !(++count % 10) { println ""; }
        }
        i = (i > 2) ? i + 2 : 3;
    }
    println "\n\n{count} such primes found.";
}
